torch/csrc/autograd/python_variable.cpp

```diff
@@ -809,8 +868,180 @@ static PyMethodDef extra_methods[] = {
+            if (obj != NULL) {
+                *(PyObject **)addr = NULL;
+                Py_DECREF(obj);
+            }
+        }
+    }
+}
+
+void THPVariable_subclass_dealloc(PyObject* self) {
+  if (THPVariable_tryResurrect((THPVariable*)self)) return;
+
+  // This is like a crappy version of subtype_dealloc.
+  // Unfortunately, we cannot directly delegate to
+  // subtype_dealloc as it will start walking the parent
+  // chain *starting with* the type of self, which will cause
+  // us to go back to our custom dealloc.
+  //
+  // We have to replicate the subtype_dealloc logic to ensure
+  // that finalizers are handled correctly
+  PyTypeObject* type = Py_TYPE(self);
+  TORCH_INTERNAL_ASSERT(type->tp_flags & Py_TPFLAGS_HEAPTYPE);
+  TORCH_INTERNAL_ASSERT(PyType_IS_GC(type), "GC types not implemented");
+
+  PyObject_GC_UnTrack(self);
+  // TODO: consider using trash can
+
+  bool has_finalizer = type->tp_finalize || type->tp_del;
+
+  if (type->tp_finalize) {
+    PyObject_GC_Track(self);
+    if (PyObject_CallFinalizerFromDealloc(self) < 0) {
+      /* Resurrected */
+      return;
+    }
+    PyObject_GC_UnTrack(self);
+  }
+
+  // base test is unnecessary as THPVariable does not set this
+  if (type->tp_weaklistoffset) {
+    PyObject_ClearWeakRefs(self);
+  }
+
+  if (type->tp_del) {
+    PyObject_GC_Track(self);
+    type->tp_del(self);
+    if (self->ob_refcnt > 0) {
+        /* Resurrected */
+        return;
+    }
+    PyObject_GC_UnTrack(self);
+  }
+
+  if (has_finalizer) {
+      /* New weakrefs could be created during the finalizer call.
+         If this occurs, clear them out without calling their
+         finalizers since they might rely on part of the object
+         being finalized that has already been destroyed. */
+      if (type->tp_weaklistoffset) {
+          /* Modeled after GET_WEAKREFS_LISTPTR() */
+          PyWeakReference **list = (PyWeakReference **) \
+              PyObject_GET_WEAKREFS_LISTPTR(self);
+          while (*list)
+              _PyWeakref_ClearRef(*list);
+      }
+  }
+
+  // Clear all slots until we get to THPVariable_dealloc
+  {
+    PyTypeObject* base = type;
+    while (base->tp_dealloc != (destructor)&THPVariable_dealloc) {
+      if (Py_SIZE(base)) {
+        clear_slots(base, self);
+      }
+      base = base->tp_base;
+      TORCH_INTERNAL_ASSERT(base);
+    }
+  }
+
+  // All Python defined classes have __dict__
+  if (C10_LIKELY(type->tp_dictoffset)) {
+      PyObject **dictptr = _PyObject_GetDictPtr(self);
+      if (dictptr != NULL) {
+          PyObject *dict = *dictptr;
+          if (dict != NULL) {
+              Py_DECREF(dict);
+              *dictptr = NULL;
+          }
+      }
+  }
+
+  // subtype_dealloc allows for this but we don't
+  TORCH_INTERNAL_ASSERT(Py_TYPE(self) == type);
+
+  // TODO: should avoid trying to resurrect again here
+  PyObject_GC_Track(self);
+  THPVariable_dealloc((THPVariable*)self);
+
+  // THPVariable is never HEAPTYPE
```

Comment:
I am confused by this comment. `THPVariableType` is not a heap type, but there are no instances of `THPVariableType` -- everything is an instance of a subtype (e.g. torch.Tensor) and those types are heap types.

In this context, I'd expect the variable `type` to **always** be a heap type.

Reply:
The comment is referencing the fact that in the `subclass_dealloc`, it tests both to see if the current type and the base type are HEAPTYPE. I'm remarking here that it's OK to skip this test on THPVariableType because it's never a heap type. 

That being said, I can see why this obscures more than illuminates, so let's replace this with an assert saying that type is a heap type!